Get OidcClient filters work with Reactive RestClient

sberyozkin opened the pull request:
Fixes #16523.

This PR introduces a `quarkus-oidc-client-reactive-filter` extension which ships a non-blocking `OidcClientReactiveFilter`. Tests check `OidcClientReactiveFilter` can acquire and refresh the token; the blocking custom filter is also tested

sberyozkin committed:
Get OidcClient filters work with Reactive RestClient

geoand commented:
Thanks @sberyozkin!

I feel we should have some documentation about this

geoand commented on extensions/oidc-client-reactive-filter/runtime/src/main/java/io/quarkus/oidc/client/reactive/filter/OidcClientRequestReactiveFilter.java:
```diff
@@ -0,0 +1,52 @@
+import io.quarkus.oidc.common.runtime.OidcConstants;
+
+@Provider
+@Singleton
```
Is this needed?

sberyozkin replied:
Just removed locally, all is fine without it

geoand commented:
BTW, can you also remove the lambdas from `AbstractTokensProducer` in another commit?

Thanks!

sberyozkin commented:
Hi @geoand 

thanks for starting reviewing,
> I feel we should have some documentation about this

Definitely, I was planning to do it before opening it for review :-). At the moment I see the post construct producer method being called during the first request which is unexpected, so I'd like to check it first - for the tests to pass I had to disable the early token acquisition in the test configuration

geoand commented:
👍🏼 

Thanks @sberyozkin

sberyozkin commented:
@geoand So with RestEasy Reactive (not sure if a reactive client has anything to do with it) it is not possible to have a `@PostConstruct` method which does a blocking call - since it is called on the IO thread.
This is how `OidcClientRequestReactiveFilter` is affected - it extends `AbstractTokensProducer` which in its `@PostConstruct` method uses `OidcClient` to acquire the tokens before the filter method is called - and it blocks and the test fails with
```
2021-04-21 14:06:08,591 ERROR [org.jbo.res.rea.com.cor.AbstractResteasyReactiveContext] (vert.x-eventloop-thread-8) Request failed: java.lang.RuntimeException: Error injecting io.quarkus.it.keycloak.ProtectedResourceServiceReactiveFilter io.quarkus.it.keycloak.FrontendResource.protectedResourceServiceReactiveFilter
	at io.quarkus.it.keycloak.FrontendResource_Bean.create(FrontendResource_Bean.zig:222)
        ....
	at io.quarkus.arc.runtime.BeanContainerImpl$1.create(BeanContainerImpl.java:35)
	at io.quarkus.resteasy.reactive.common.runtime.ArcBeanFactory.createInstance(ArcBeanFactory.java:25)
        ....
	at org.jboss.resteasy.reactive.server.vertx.ResteasyReactiveVertxHandler.handle(ResteasyReactiveVertxHandler.java:7)
	at io.vertx.ext.web.impl.RouteState.handleContext(RouteState.java:1129)
	............
Caused by: java.lang.IllegalStateException: The current thread cannot be blocked: vert.x-eventloop-thread-8
	at io.smallrye.mutiny.operators.uni.UniBlockingAwait.await(UniBlockingAwait.java:30)
        ...
	at io.quarkus.oidc.client.runtime.TokensHelper.initTokens(TokensHelper.java:20)
	at io.quarkus.oidc.client.runtime.AbstractTokensProducer.initTokens(AbstractTokensProducer.java:34)
	at io.quarkus.oidc.client.reactive.filter.OidcClientRequestReactiveFilter.initTokens(OidcClientRequestReactiveFilter.java:30)
	at io.quarkus.oidc.client.reactive.filter.OidcClientRequestReactiveFilter_Bean.create(OidcClientRequestReactiveFilter_Bean.zig:298)
	...
	at io.quarkus.arc.impl.AbstractSharedContext.get(AbstractSharedContext.java:26)
	at io.quarkus.oidc.client.reactive.filter.OidcClientRequestReactiveFilter_Bean.get(OidcClientRequestReactiveFilter_Bean.zig:346)    
        .......
	at io.quarkus.arc.impl.ArcContainerImpl.instance(ArcContainerImpl.java:222)
	at io.quarkus.rest.client.reactive.runtime.BeanGrabber.getBeanIfDefined(BeanGrabber.java:9)
	at io.quarkus.it.keycloak.ProtectedResourceServiceReactiveFilter$$QuarkusRestClientInterface.<init>(ProtectedResourceServiceReactiveFilter$$QuarkusRestClientInterface.zig:36)
	at io.quarkus.it.keycloak.ProtectedResourceServiceReactiveFilter$$QuarkusRestClientInterfaceCreator.apply(ProtectedResourceServiceReactiveFilter$$QuarkusRestClientInterfaceCreator.zig:34)
	at org.jboss.resteasy.reactive.client.impl.ClientProxies.get(ClientProxies.java:28)
	at org.jboss.resteasy.reactive.client.impl.WebTargetImpl.proxy(WebTargetImpl.java:378)
	at io.quarkus.rest.client.reactive.runtime.RestClientBuilderImpl.build(RestClientBuilderImpl.java:261)
	at io.quarkus.rest.client.reactive.runtime.RestClientCDIDelegateBuilder.build(RestClientCDIDelegateBuilder.java:77)
	at io.quarkus.rest.client.reactive.runtime.RestClientCDIDelegateBuilder.createDelegate(RestClientCDIDelegateBuilder.java:59)
	at io.quarkus.it.keycloak.ProtectedResourceServiceReactiveFilter$$CDIWrapper.<init>(ProtectedResourceServiceReactiveFilter$$CDIWrapper.zig:26)
	at io.quarkus.it.keycloak.ProtectedResourceServiceReactiveFilter$$CDIWrapper_ClientProxy.<init>(ProtectedResourceServiceReactiveFilter$$CDIWrapper_ClientProxy.zig:28)
	at io.quarkus.it.keycloak.ProtectedResourceServiceReactiveFilter$$CDIWrapper_Bean.proxy(ProtectedResourceServiceReactiveFilter$$CDIWrapper_Bean.zig:40)
	at io.quarkus.it.keycloak.ProtectedResourceServiceReactiveFilter$$CDIWrapper_Bean.get(ProtectedResourceServiceReactiveFilter$$CDIWrapper_Bean.zig:206)
	at io.quarkus.it.keycloak.ProtectedResourceServiceReactiveFilter$$CDIWrapper_Bean.get(ProtectedResourceServiceReactiveFilter$$CDIWrapper_Bean.zig:222)
	at io.quarkus.it.keycloak.FrontendResource_Bean.create(FrontendResource_Bean.zig:205)
```

This is why I have to set `quarkus.oidc-client.early-token-acquisition=false` to avoid having that call in `PostConstract` - which means the initial token acquisition has to be shifted to the moment the first filter request is done.
Not a problem for this PR as such - I'll doc this restriction.

Do you think it makes sense to open a new issue to track this particular problem - or is it the inherent restriction - don't do blocking call even at the bean initialization time when RestEasy Reactive is used ? (This is when DB connections can be made, etc)  ?

sberyozkin commented:
Actually - I probably can update that code to run on the blocking executor if no Blocking is allowed

geoand commented:
@sberyozkin so indeed filter should never block - even when those are being initialized, because that initialization happens on the IO thread.

We really need to find a way to make evething the filter does non-blocking

geoand commented:
Can't we have the filter contain state that populated the first time?

sberyozkin commented:
@geoand 

> Can't we have the filter contain state that populated the first time the filter is executed?

Yes, this is what `quarkus.oidc-client.early-token-acquisition=false` does.
Ideally though I'd like to have an option for the initialization to be done before the filter is executed - as it can be executed long after the server started - thus delaying the initialization cost till the first filter call may not be ideal - though in case of reactive OidcClient filters the longer term perf savings would definitely outweigh the cost of the init delay :-)

geoand commented:
>though in case of reactive OidcClient filters the longer term perf savings would definitely outweigh the cost of the init delay :-)

My thoughts exactly :)

sberyozkin commented:
@geoand This is now ready for review. CC-ing to @pedroigor as well - but nothing is changing at the OIDC client level per se :-), only a reactive filter is introduced
thanks

geoand commented on docs/src/main/asciidoc/security-openid-connect-client.adoc:
```diff
@@ -133,6 +133,31 @@ public interface ProtectedResourceService {
+
+It works similary to the way `OidcClientRequestFilter` (described in the previous section) does - it uses `OidcClient` to acquire the access token, refresh it if needed, and set it as an HTTP `Authorization` `Bearer` scheme value. The difference is that it works with link:rest-client-reactive[Reactive RestClient] and implements a non-blocking client filter which does not block the current IO thre ...
+
+At the moment this filter expects that the token acquisition is delayed with `quarkus.oidc-client.early-tokens-acquisition=false` and it can only be registered with `org.eclipse.microprofile.rest.client.annotation.RegisterProvider` annotation:
```
So a user needs to set `quarkus.oidc-client.early-tokens-acquisition=false` for this to work properly?

Can't we just ignore this configuration option entirely as it really is not safe to use?

sberyozkin replied:
@geoand I'll try to override the init method - as it works by default for OidcClientRequestFilter

geoand commented on extensions/oidc-client-reactive-filter/runtime/src/main/java/io/quarkus/oidc/client/reactive/filter/OidcClientRequestReactiveFilter.java:
```diff
@@ -0,0 +1,56 @@
+
+    protected void initTokens() {
+        if (earlyTokenAcquisition) {
+            LOG.debug("Token acquisition will be delayed until this filter is executed to avoid blocking an IO thread");
```
👍🏼

geoand replied:
Perhaps also mentione explicitly that the config option that will be ignored to make it actionable for the user?

sberyozkin replied:
I was hoping you'd like this debug message :-)

sberyozkin commented:
There was some pom verification error during the previous build - and checked everything and could not spot anything that could've caused it - the only suspect was it was also returning `Feature.OIDC_CLIENT_FILTER`, same as `quarkus-oidc-client-filter` so I updated that to return  `Feature.OIDC_CLIENT_REACTIVE_FILTER`

sberyozkin commented:
Hmm, seeing
```
2021-04-21T15:27:41.4880813Z [0;31mError:[0m Dependencies to extension artifacts are outdated!
2021-04-21T15:27:41.4882801Z [0;31mError:[0m Run ./update-extension-dependencies.sh and add the modified pom.xml files to your commit.
2021-04-21T15:27:41.4884749Z [0;31mError:[0m Diff is:
2021-04-21T15:27:41.5050042Z diff --git a/devtools/bom-descriptor-json/pom.xml b/devtools/bom-descriptor-json/pom.xml
2021-04-21T15:27:41.5051246Z index 7922a8e6..d9d2d212 100644
2021-04-21T15:27:41.5052256Z --- a/devtools/bom-descriptor-json/pom.xml
2021-04-21T15:27:41.5059952Z +++ b/devtools/bom-descriptor-json/pom.xml
2021-04-21T15:27:41.5061045Z @@ -1566,6 +1566,19 @@
2021-04-21T15:27:41.5061576Z                          </exclusion>
2021-04-21T15:27:41.5061978Z                      </exclusions>
2021-04-21T15:27:41.5062367Z                  </dependency>
2021-04-21T15:27:41.5062768Z +                <dependency>
2021-04-21T15:27:41.5063395Z +                    <groupId>io.quarkus</groupId>
2021-04-21T15:27:41.5065034Z +                    <artifactId>quarkus-oidc-client-reactive-filter</artifactId>
2021-04-21T15:27:41.5066052Z +                    <version>${project.version}</version>
2021-04-21T15:27:41.5066573Z +                    <type>pom</type>
2021-04-21T15:27:41.5067001Z +                    <scope>test</scope>
2021-04-21T15:27:41.5067409Z +                    <exclusions>
2021-04-21T15:27:41.5067818Z +                        <exclusion>
2021-04-21T15:27:41.5068235Z +                            <groupId>*</groupId>
2021-04-21T15:27:41.5068720Z +                            <artifactId>*</artifactId>
2021-04-21T15:27:41.5069173Z +                        </exclusion>
2021-04-21T15:27:41.5069584Z +                    </exclusions>
2021-04-21T15:27:41.5070001Z +                </dependency>
2021-04-21T15:27:41.5070407Z                  <dependency>
2021-04-21T15:27:41.5070881Z                      <groupId>io.quarkus</groupId>
2021-04-21T15:27:41.5071749Z                      <artifactId>quarkus-oidc-common</artifactId>
2021-04-21T15:27:41.5072738Z diff --git a/docs/pom.xml b/docs/pom.xml
2021-04-21T15:27:41.5074833Z index 2fb17020..86b4460b 100644
2021-04-21T15:27:41.5075575Z --- a/docs/pom.xml
2021-04-21T15:27:41.5076281Z +++ b/docs/pom.xml
2021-04-21T15:27:41.5076779Z @@ -1527,6 +1527,19 @@
2021-04-21T15:27:41.5077146Z                  </exclusion>
2021-04-21T15:27:41.5077533Z              </exclusions>
2021-04-21T15:27:41.5077936Z          </dependency>
2021-04-21T15:27:41.5078325Z +        <dependency>
2021-04-21T15:27:41.5078785Z +            <groupId>io.quarkus</groupId>
2021-04-21T15:27:41.5080011Z +            <artifactId>quarkus-oidc-client-reactive-filter-deployment</artifactId>
2021-04-21T15:27:41.5081100Z +            <version>${project.version}</version>
2021-04-21T15:27:41.5081610Z +            <type>pom</type>
2021-04-21T15:27:41.5082010Z +            <scope>test</scope>
2021-04-21T15:27:41.5082421Z +            <exclusions>
2021-04-21T15:27:41.5082802Z +                <exclusion>
2021-04-21T15:27:41.5083227Z +                    <groupId>*</groupId>
2021-04-21T15:27:41.5083706Z +                    <artifactId>*</artifactId>
2021-04-21T15:27:41.5084159Z +                </exclusion>
2021-04-21T15:27:41.5084544Z +            </exclusions>
2021-04-21T15:27:41.5084948Z +        </dependency>
2021-04-21T15:27:41.5085353Z          <dependency>
2021-04-21T15:27:41.5085800Z              <groupId>io.quarkus</groupId>
2021-04-21T15:27:41.5086766Z              <artifactId>quarkus-oidc-common-deployment</artifactId>
2021-04-21T15:27:41.5100259Z ##[error]Process completed with exit code 1.
```
OK, just did what it says, `Run ./update-extension-dependencies.sh and add the modified pom.xml files to your commit.` :-)

sberyozkin commented:
Not sure what the problem with the Windows build, apparently it is
```
[INFO] --- quarkus-maven-plugin:999-SNAPSHOT:build (default) @ quarkus-integration-test-jpa-without-entity ---
[INFO] [org.jboss.threads] JBoss Threads version 3.2.0.Final
Error: The operation was canceled.

```

geoand commented:
> Not sure what the problem with the Windows build, apparently it is
> 
> ```
> [INFO] --- quarkus-maven-plugin:999-SNAPSHOT:build (default) @ quarkus-integration-test-jpa-without-entity ---
> [INFO] [org.jboss.threads] JBoss Threads version 3.2.0.Final
> Error: The operation was canceled.
> ```

I don't see how it could be related to the PR

geoand commented:
I don't think you have added the new test to the native builds for CI, or did I miss it?

sberyozkin commented:
@geoand good catch, will do; also I'd like to re-work the test which depends on `@Blocking` - I just started with that test to check the injection problem; so I'll replace the custom blocking filter with a custom non-blocking filter and update the docs

geoand commented:
Makes sense

sberyozkin commented:
I've reworked the test but stopped short of updating the docs with the example of a custom filter because it is a copy and paste of the one from the new extension. I thought for a sec I could show the `Tokens` producer - but that one is blocking so it won't do.

geoand reviewed (approved):
LGTM!

sberyozkin commented:
@geoand Thanks for making it happen - I just copied and pasted your test code :-)

geoand commented:
Hahaha :)

Glad it worked!

sberyozkin commented:
Hmm, the custom filter test was passing for me in JVM/native test

sberyozkin commented:
Hi @geoand and @michalszynkiewicz I've rebased and rerun this test and it is all green locally - FYI, I was seeing this issue,
```
Expected status code <200> but was <204>.
```
when I was starting to work on the test - the problem was, I thought, that `ProtectedResource` endpoint and MP REST clients had no `@Produces("text/plain")` as soon as I added them it started working.
What I'll try and add `Accept: text/plain` and see if it helps

geoand commented:
That's one for @michalszynkiewicz I think :)

sberyozkin commented:
I've added `Accept: text/plain` - no idea if it will help in CI though - but it passes locally. The other 'suspect' is - some caching is involved the the client request level ? that may also cause `204`

sberyozkin commented:
OK, I confirm I've no idea what is going on, I nearly regret I added a 2nd custom filter test :-) which is not even part of the new extension - but something is not right...

geoand commented:
I'll have a look tomorrow or Monday

sberyozkin commented:
Hey @geoand sorry about this test failing - I'll try to isolate the problem as wel

geoand commented:
No worries ☺️

sberyozkin commented:
I'll create a branch on top of this one and start tweaking to see how CI reacts - locally all works fine

sberyozkin commented:
I've started with disabling the test depending on the filter shipped with `quarkus-oidc-client-reactive-filter` and only kept a single test which is failing here and which depends on a custom filter included with in the test module, [it works](https://pipelines.actions.githubusercontent.com/vfSsPy88KnBzCO8MZariDJqIaxlYhLVSkk7Gvtuq0Ay08FAQSp/_apis/pipelines/1/runs/53357/signedlogcontent/45?urlExpires=2021-04-23T13%3A27%3A56.0242430Z&urlSigningMethod=HMACV1&urlSignature=SyR4ErOk%2FCWOLuq7Wi4WleZK6%2BDWKRH5iZMIJC4bHho%3D). It took nearly half a day for the build in my fork to complete though :-)
Not sure yet what it might mean.

sberyozkin commented:
Next step is re-enable both tests and add a few log message checking what has been invoked etc

sberyozkin commented:
Running both tests (with a custom filter and with the one provided by the extension works passes in my fork)

sberyozkin commented:
OK, I've confirmed it is failing in CI if all 3 tests are executed as in this PR... There could many options why it fails, so will take a bit longer for this PR to be merged :-). It can be a problem with OidcClient's AbstractTokensProducer which the reactive client is exposing or it can be a case of the reactive client mixing the filters somehow - we'll see :-)

geoand commented:
I'll try and take a look tomorrow

On Sun, Apr 25, 2021, 13:09 sberyozkin ***@***.***> wrote:

> OK, I've confirmed it is failing in CI if all 3 tests are executed as in
> this PR... There could many options why it fails, so will take a bit longer
> for this PR to be merged :-). It can be a problem with OidcClient's
> AbstractTokensProducer which the reactive client is exposing or it can be a
> case of the reactive client mixing the filters somehow - we'll see :-)
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/quarkusio/quarkus/pull/16658#issuecomment-826296972>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABBMDP4WGC32GQEMHBN3HBTTKPS3ZANCNFSM43IHDEKA>
> .
>

sberyozkin commented:
Added some logging on my branch to see how the failing test is progressing

geoand commented:
I can't reproduce the issue locally

geoand commented on integration-tests/oidc-client-reactive/src/main/java/io/quarkus/it/keycloak/OidcClientRequestCustomFilter.java:
```diff
@@ -0,0 +1,57 @@
+            @Override
+            public void accept(Tokens tokens) {
+                requestContext.resume();
+                requestContext.getHeaders().add(HttpHeaders.AUTHORIZATION, BEARER_SCHEME_WITH_SPACE + tokens.getAccessToken());
```
Actually, `resume()` should be the last thing to do as there is a small chance of a race condition. 

Same for error handler

geoand commented on extensions/oidc-client-reactive-filter/runtime/src/main/java/io/quarkus/oidc/client/reactive/filter/OidcClientRequestReactiveFilter.java:
```diff
@@ -0,0 +1,56 @@
+            @Override
+            public void accept(Tokens tokens) {
+                requestContext.resume();
+                requestContext.getHeaders().add(HttpHeaders.AUTHORIZATION, BEARER_SCHEME_WITH_SPACE + tokens.getAccessToken());
```
Same comment as in `io.quarkus.it.keycloak.OidcClientRequestCustomFilter`

geoand commented:
@sberyozkin I added a comment that _could_ be the cause of the problem

sberyozkin commented:
Hey @geoand 
I'm sorry, just seeing these comments now, came here to report the tests passed in [my fork}(https://pipelines.actions.githubusercontent.com/vfSsPy88KnBzCO8MZariDJqIaxlYhLVSkk7Gvtuq0Ay08FAQSp/_apis/pipelines/1/runs/53376/signedlogcontent/44?urlExpires=2021-04-26T17%3A42%3A52.8402296Z&urlSigningMethod=HMACV1&urlSignature=pAzRdYzqNXlUO02g1JuVCDVhkuOmc3Acs%2Bif221D5u0%3D)

OK. let me move `resume` to the end, thanks for checking it

sberyozkin commented:
Same docker problem, will retry tomorrow

gsmet commented:
I think this one will be in the same ballpark as all the other REST Client Reactive PRs to backport. Either someone works on backporting all of them and fixing conflicts or they won't be backported and will wait for 2.0.

sberyozkin commented:
Hi @gsmet I'm fine with waiting till 2.0, @geoand hope you agree, let me remove the backport label

geoand commented:
> Hi @gsmet I'm fine with waiting till 2.0, @geoand hope you agree, let me remove the backport label

Sure thing
